### PR TITLE
Feature/quiz functionality

### LINKED ISSUE #500: Implement Basic Quiz Functionality
Documentation: [Creating Apps With Quizzes](https://software.sil.org/downloads/r/scriptureappbuilder/Scripture-App-Builder-11-Creating-Quizzes.pdf)
Presentation: [Introduction to Scripture App Builder](https://docs.google.com/presentation/d/11FFg8w0K83lcCsbavXPCUdeVWOAYOJca/edit?usp=sharing&ouid=116747902156680384840&rtpof=true&sd=true)
* See section on quizzes

Quizzes may be configured in SAB as a book type.  Below is an example of a quiz with different format questions:
\id QU1
\quiz

\qu Who was the father of Jacob?
\qu FatherQu.mp3
\aw Abraham
\aw Abraham.mp3
\ar Isaac
\ar Isaac.mp3
\aw Joseph
\aw Joseph.mp3
\aw Esau
\aw Esau.mp3

\qu Where was Jesus born?
\ar Bethlehem
\aw Nazareth
\aw Jerusalem
\aw Bethany
\ae Jesus was born in Bethlehem

\qu Where are the dates?
\aw apples.jpg
\aw bitter-tomato.jpg
\aw cucumbers.jpg
\ar dates.jpg

\qu What is this?
\qu bitter-tomato.jpg
\aw Apples
\ar Ngoyo
\aw Cucumbers
\aw Dates

\sc 4 Perfect!
\sc 3 Almost
\sc 0 Oh dear!

\pm 2

qu is the question.  The last question shows one with text and an image
aw is a wrong answer.  This may be text or an image. 
ar is the correct answer which may also be text or image
ae is text that displays if the answer selected is incorrect

ac is the column count for the number of columns to use to display the answer choices.  By default, it is set to 2 if the question has an image and more than two answers or if any of the answers to the has an image.  Otherwise the default is 1.

sb Quiz Score page: line at top of page.  Defaults to string id Quiz_Score_Page_Message_Before if not set in quiz ("You Scored")

sa Quiz Score page: line at bottom.  Defaults to string id Quiz_Score_Page_Message_After if not set in quiz ("out of %n% questions").  When this message is used, any %n% in the string is replaced by the total number of questions in the quiz.

sc Score commentary.  If the score equals the value, the text is shown below the score

ra Right answer audio file to play (defaults to quiz-right-answer.mp3 if not set for quiz)
wa Wrong answer audio file to play (defaults to quiz-wrong-answer.mp3 if not set for quiz)
pm Score at or above this value is a passing score

The first question shows a question with audio clips for the question and audio.  When the question is displayed, the audio for the question is read then each answer is highlighted and its audio played.  The user can press an answer while the audio is playing which terminates the audio play of the answers.  

Image

The second question shows a simple question that displays an explanation if the answer is incorrect
quizIncorrect

The third question has images as the answers
pictureAnswerCorrect

The fourth question has an image as the question
QuizPictureQuestion

And finally a score screen:
Score screen

### PR DESCRIPTION
**Pull request by aidanpscott**
Fixes #500

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # appbuilder-pwa
 
-A template for building a Progressive Web App with [Scripture App Builder](https://softare.sil.org/scriptureappbuilder).
+A template for building a Progressive Web App with [Scripture App Builder](https://software.sil.org/scriptureappbuilder).
```
Comment (chrisvire) [inline README.md:3]: Why is this change included?  This is from a previous PR.

File: README.md [hunk 2]
```diff
@@ -55,3 +55,22 @@ Run `npm test` to run created test files.
...
+```bash
+npm install
+npm run convert
+cd data-sandbox
+npm install
+cd ..
+npm run sandbox
+```
+
+Open a browser to http://localhost:2468/graphql to query the Scripture data using GraphQL.
```
Comment (chrisvire) [inline README.md:76]: Why is this change included?  This is from a previous PR.

**Review by chrisvire — CHANGES_REQUESTED** [review]
This is a really good start.  I have submitted some fixed to the PR so you will need to `git pull -r` before making more changes.

There seems to be commits in this PR that are already in main.  

Things that are broken:
* question that is a picture is too big on desktop.
* when answering with picture, it should highlight the picture (in green and red) after selecting an answer (similar to changing the color of the answer bubbles).

**Review by chrisvire — CHANGES_REQUESTED** [review]
It is looking pretty good.  There seems to be a change from previous behavior.  It should:

When an answer is selected, it should:
1. highlight the answer selected as right or wrong and play the sound
2. if the answer was wrong, wait a second and then highlight the correct answer

File: src/lib/components/BookSelector.svelte [hunk 3]
```diff
@@ -113,24 +132,36 @@ The navbar component.
         nextRef.reset();
     }
 
-    /**list of books in current docSet*/
+    /**list of books, quizzes, and quiz groups in current docSet*/
     $: books = $refs.catalog.documents;
+    $: quizzes = $refs.catalog.quizzes;
```
Comment (chrisvire) [inline src/lib/components/BookSelector.svelte:137]: We are not using this, correct?

**Comment by chrisvire** [comment]
Sometimes I am getting this:

image

**Comment by chrisvire** [comment]
If I start quiz and then while it is playing audio and I navigate way (i.e. hit the back arrow), the audio is still playing.

File: src/lib/components/BookSelector.svelte (hunk 3 above)
Comment (aidanpscott) [reply src/lib/components/BookSelector.svelte:137]: This is correct. I thought I deleted that.

**Review by chrisvire — CHANGES_REQUESTED** [review]
You need to include implementation of answer (both text and audio).

**Review by chrisvire — CHANGES_REQUESTED** [review]
This requirement has not been implemented:

> ac is the column count for the number of columns to use to display the answer choices. By default, it is set to 2 if the question has an image and more than two answers or if any of the answers to the has an image. Otherwise the default is 1.

File: convert/convertBooks.ts [hunk 4]
```diff
@@ -201,8 +201,9 @@ export async function convertBooks(
                             'quizzes',
                             book.id + '.json'
                         ),
-                        content: JSON.stringify(convertQuizBook(context, book))
+                        content: JSON.stringify(convertQuizBook(context, book), null, 2)
                     });
+                    process.stdout.write(` ${book.id}`);
```
Comment (davidmoore1) [inline convert/convertBooks.ts:206]: Can't put a comment on line 255, but it crashes there if the project only contains quizzes (At least that's what I think did it.  A project with two quizzes in the main book collection and that is all).  Entries contains a single entry with no docSets, so docSets[0] didn't exist so it blew up referencing its id.  A project with only quizzes is a valid combination and can't crash on startup.

File: src/lib/components/HistoryCard.svelte [hunk 5]
```diff
@@ -22,21 +23,29 @@ TODO:
...
 
-<!-- history cards are alway LTR with the reference following the text direction -->
-<div class="history-item-block dy-card w-100 bg-base-100 shadow-lg my-4" style:direction="ltr">
-    <!-- svelte-ignore a11y-click-events-have-key-events -->
-    <a
-        style="text-decoration:none;"
-        href="{base}/"
-        on:click={() =>
+    function onHistoryClick() {
+        if (history.url) {
```
Comment (davidmoore1) [inline src/lib/components/HistoryCard.svelte:28]: This is all good.  I hadn't even thought about the fact that the quiz should appear in history.

**Review by davidmoore1 — CHANGES_REQUESTED** [review]

**Review by davidmoore1 — APPROVED** [review]

File: convert/convertBooks.ts (hunk 4 above)
Comment (chrisvire) [reply convert/convertBooks.ts:206]: There is a separate PR for this: #580.  It needs closer review.

File: src/routes/quiz/[collection]/[id]/+page.svelte [hunk 6]
```diff
@@ -330,25 +330,25 @@
                                 {#if currentQuizQuestion.image}
                                     <!-- svelte-ignore a11y-missing-attribute -->
                                     <img
-                                        class="quiz-question-image h-40"
```
Comment (chrisvire) [inline src/routes/quiz/[collection]/[id]/+page.svelte]: I saw that you removed h-40 from quiz-question-image.  That still needs to be there.  The android app hardcodes the height of the question image to 160px (h-40 equivalent).

**Review by chrisvire — CHANGES_REQUESTED** [review]
Almost there. Please make these changes.

1.  Remove extra commit for next issue (it needs its own branch/PR).
2. Re-add h-40 to quiz-question-image.

**Review by chrisvire — APPROVED** [review]
Looks good to go now.  Thanks for all the hard work and persisting!
